examples/ecs/observers.rs

```diff
@@ -0,0 +1,199 @@
+//! Demonstrates how to observe life-cycle triggers as well as define custom ones.
+
+use bevy::{
+    prelude::*,
+    utils::{HashMap, HashSet},
+};
+
+fn main() {
```

Comment:
My big architectural thought is that it feels like observers can be implemented in "userspace" rather than adding them as a new storage. This implements essentially the same behavior with _significantly_ simplified internals + dataflow.

```rust
fn main() {
    App::new()
        .init_resource::<ObserverHandlers<Foo>>()
        .add_plugins(DefaultPlugins)
        .add_systems(Startup, setup)
        .run();
}

struct Foo(usize);

fn setup(mut commands: Commands, mut handlers: ResMut<ObserverHandlers<Foo>>) {
    let e1 = commands
        .spawn(On::new(|observer: Observer<Foo>| {
            println!("Entity {:?} Foo {}", observer.source(), observer.data().0);
        }))
        .id();

    commands.add(SendEntityEvent {
        value: Foo(10),
        entities: [e1],
    });

    handlers.add(|observer: Observer<Foo>| println!("Foo {}", observer.data().0));
    commands.add(SendEvent { value: Foo(20) });
}

#[derive(Default)]
struct On<T: 'static> {
    // TODO: Make this a SmallVec with 1 capacity
    observers: Vec<BoxedObserverSystem<T>>,
}

impl<T: 'static> Component for On<T> {
    const STORAGE_TYPE: bevy::ecs::component::StorageType = StorageType::SparseSet;
    fn register_component_hooks(hooks: &mut ComponentHooks) {
        hooks.on_add(|mut world, entity, _| {
            let observers = std::mem::take(
                &mut world
                    .entity_mut(entity)
                    .get_mut::<On<T>>()
                    .unwrap()
                    .observers,
            );
            let mut handlers = world.resource_mut::<ObserverHandlers<T>>();
            handlers.entity_handlers.insert(entity, observers);
        });
        hooks.on_remove(|mut world, entity, _| {
            let mut handlers = world.resource_mut::<ObserverHandlers<T>>();
            handlers.entity_handlers.remove(&entity);
        });
    }
}

impl<T> On<T> {
    pub fn new<M>(system: impl IntoObserverSystem<T, (), M>) -> Self {
        let system = IntoObserverSystem::into_system(system);
        let system: BoxedObserverSystem<T> = Box::new(system);
        Self {
            observers: vec![system],
        }
    }
}

#[derive(Resource)]
struct ObserverHandlers<T: 'static> {
    handlers: Vec<BoxedObserverSystem<T>>,
    entity_handlers: HashMap<Entity, Vec<BoxedObserverSystem<T>>>,
}

impl<T> ObserverHandlers<T> {
    pub fn add<M>(&mut self, system: impl IntoObserverSystem<T, (), M>) {
        let system = IntoObserverSystem::into_system(system);
        let system: BoxedObserverSystem<T> = Box::new(system);
        self.handlers.push(system);
    }
}

impl<T> Default for ObserverHandlers<T> {
    fn default() -> Self {
        Self {
            handlers: Default::default(),
            entity_handlers: Default::default(),
        }
    }
}

struct SendEntityEvent<T: Send + 'static, E: IntoIterator<Item = Entity>> {
    value: T,
    entities: E,
}

impl<T: Send + Sync + 'static, E: IntoIterator<Item = Entity> + Send + 'static> Command
    for SendEntityEvent<T, E>
{
    #[allow(unsafe_code)]
    fn apply(mut self, world: &mut World) {
        world.resource_scope(|world, mut events: Mut<ObserverHandlers<T>>| {
            let events = &mut *events;
            for (entity, handlers) in &mut events.entity_handlers {
                for handler in handlers {
                    let observer_trigger = ObserverTrigger {
                        observer: *entity,
                        source: *entity,
                        trigger: ComponentId::new(10),
                    };
                    // TODO: for simplicity I'm initializing here. In practice this should only be done once
                    handler.initialize(world);
                    handler.run(
                        Observer::new(
                            unsafe { std::mem::transmute(&mut self.value) },
                            observer_trigger,
                        ),
                        world,
                    );
                }
            }
        });
    }
}

struct SendEvent<T: Send + 'static> {
    value: T,
}

impl<T: Send + Sync + 'static> Command for SendEvent<T> {
    fn apply(mut self, world: &mut World) {
        world.resource_scope(|world, mut events: Mut<ObserverHandlers<T>>| {
            let events = &mut *events;
            for handler in &mut events.handlers {
                let observer_trigger = ObserverTrigger {
                    observer: Entity::PLACEHOLDER,
                    source: Entity::PLACEHOLDER,
                    trigger: ComponentId::new(10),
                };
                // TODO: for simplicity I'm initializing here. In practice this should only be done once
                handler.initialize(world);
                handler.run(
                    Observer::new(
                        unsafe { std::mem::transmute(&mut self.value) },
                        observer_trigger,
                    ),
                    world,
                );
            }
        });
    }
}
```

So I think the question becomes "why not do this"?

Reply:
(disclaimer: the code above is not production quality and takes a number of shortcuts, but it does compile and work!)

Reply:
A thought for the code above: to ensure we initialize systems ahead of time we could just require going through world methods / commands to register handlers.

Reply:
This works fine for user defined events, I don't believe it maps well to usage for the internal events such as `OnInsert` and `OnRemove` where you would be adding significant additional indirection to the hot path by having to look up each resource. There's also the additional overhead of using a command but that could probably be avoided. 

Additionally there needs to be room in the design for listening to events targeting specific components which is key for query observers (and several user use cases). The main impetus for observers is to allow updating query caches so invocations need to have minimal overhead otherwise we will suffer heavily when archetype fragmentation increases. To be fair this implementation isn't *heavily* optimised, but it leaves room for that since we have so much control over the internals.

Reply:
You also need a dynamic API, which makes the implementation much less trivial than this example.

Reply:
> This works fine for user defined events, I don't believe it maps well to usage for the internal events such as OnInsert and OnRemove where you would be adding significant additional indirection to the hot path by having to look up each resource

Yeah I agree on this / it would need an equivalent fast path to be preferential to the current impl. We could investigate fast path approaches (ex: cache pointers to those event resources as they _should_ be stable in the context of evaluating these handlers, don't use resources for the "built in" events, etc)

> There's also the additional overhead of using a command but that could probably be avoided.

Yeah we would just factor out the event evaluation code / apply it directly.

> Additionally there needs to be room in the design for listening to events targeting specific components which is key for query observers (and several user use cases)

I think this could be accomplished with a `ComponentId->ObserverStorage` resource and a new `SendComponentEvent ` equivalent to the code above.

> You also need a dynamic API, which makes the implementation much less trivial than this example.

Same as the above. Although a dynamic API would require additional changes (just like it would in this PR's impl).

That being said, I'm not _really_ opposed to the internalization of the observer storage (I'm just always on the hunt for concept unification). My biggest hangup right now is the indirection between entities and their observers. It seems like an entity observer should be directly correlated to (and "owned by") the entity instead of having a secondary "observer entity" that correlates to the actual entity. This feels like unnecessary bookkeeping, introducing both code complexity and overhead.

Ditto for "entity-less / component-less observers". Why introduce the entity when we could just directly insert the observer into a `Trigger -> Vec<Observer>` map stored on the World?

Reply:
The purpose of the entity is so the observers can be trivially despawned, queried for etc. All the same reasons people want systems as entities. Once we have full relations having associated entities will be much more normalized, configurable and have improved UX. 

Another reason is to allow attaching observers to entities you don't "own" i.e. observing some entity managed by an external crate that exposes triggers that you can observe. You can be confident they will be cleaned up automatically if the target is despawned, but are also able to despawn it yourself if you're no longer interested in observing the event. On the whole it's just much more dynamic then introducing a map you either have to manually update or introducing some secondary id to identify observers.

Reply:
Observers being components is a huge loss as it disallows adding multiple for a single trigger type, either because you have a dynamic scenario where you want that or because an external crate uses a trigger of that type already internally. If they can't be components you still want some way to refer to them, IMO it should be `Entity`, anything else is way less powerful.

Reply:
> The purpose of the entity is so the observers can be trivially despawned, queried for etc. All the same reasons people want systems as entities. Once we have full relations having associated entities will be much more normalized, configurable and have improved UX.
> Another reason is to allow attaching observers to entities you don't "own" i.e. observing some entity managed by an external crate that exposes triggers that you can observe. You can be confident they will be cleaned up automatically if the target is despawned, but are also able to despawn it yourself if you're no longer interested in observing the trigger. On the whole it's just much more dynamic then introducing a map you either have to manually update or introducing some secondary id to identify observers.

Ok I do see the appeal of the separate entity approach from an "observer management" perspective. Being able manually manage the lifetime of an observer seems valuable. However in the current form that doesn't seem particularly viable as we've fully abstracted over the "entity-ness" / introduced additional private internals. You can't spawn an observer the "normal way". This also means that we'd need special syntax to define observers in scenes, which gives me pause.

> Observers being components is a huge loss as it disallows adding multiple for a single trigger type, either because you have a dynamic scenario where you want that or because an external crate uses a trigger of that type already internally.

In practice I don't think this _needs_ to be true because in my impl above we consume the vec of observers immediately on spawn. A separate party trying to also observe a specific entity would be free to _also_ insert `On<Trigger>` on the entity and it would overwrite the empty list (and the `insert` hook could be used to pick up the new observers). That _does_ mean that these observers are munged together, meaning they could not be individually removed. And it does also seem janky from a dataflow perspective. I do think the entity approach is cleaner when it comes to tracking additions and removals.

I've been experimenting with a new design sketch that embraces the "separate entities" model, while embracing the normal Bevy data model and simplifying dataflow a bit.